### DISCUSSION TITLE
Multi-Threaded Nested Model Support

### DISCUSSION THREAD

**Discussion by Areopagitics** [slint-ui/slint#6735] [Q&A]
I currently have a property `in-out property <[{file:File,rows:[Row]}]> all_rows;`  that works perfectly fine when when I update it using `ui.set_all_rows()`, but if I make it multi-threaded it returns this error, and this is regardless if I use a Mutex or simply using `as_any().downcast_ref.`

`std::rc::Rc<(dyn Model<Data = slint_generatedAppWindow::Row> + 'static)> cannot be sent between threads safely`

Here's the code block using Rust:

```
ui_handle.upgrade_in_event_loop(move |h|{
                    let model_handle: ModelRc<(File, ModelRc<Row>)> = h.get_all_findings();
                    let model = model_handle.as_any().downcast_ref::<VecModel<(File, ModelRc<Row>)>>().unwrap();
                    model.push(file_rows);
                });

```

Would this be easy to implement?

**Comment by ogoffart** [comment]
The model is indeed not Send and I don't think it is possible to change that, because the view need to listen to change to the model to update itself, and this is really not something that can be done if the model changes in a thread.

What should be done is that update to the model need to be done in the main thread.
So instead of moving the model to the event loop, you should move the data, and copy it into the model in the main thread.
